src/app/tap.rs

```diff
@@ -0,0 +1,99 @@
+use super::identity;
+use futures::{future, Future};
+use logging;
+use proxy;
+use std::error;
+use svc;
+use tokio::executor;
+use tower_grpc as grpc;
+use transport::{
+    tls::{self, HasPeerIdentity},
+    Listen,
+};
+use Conditional;
+
+macro_rules! tap_server {
```

Comment:
honest question: are these macros worth it?  they're not very large and they're only invoked twice.

another honest question: why do these need to be macros rather than functions?

Reply:
They may not be worth it. They do save some lines of code, but as a reader it does introduce some complexity in figuring out what exactly is going on.

They are macros, because as functions I felt like I was spending more time making the types generic enough to satisfy both call sites. They became more complex than I thought they needed to be.

I could remove these or take another go at making them functions, but either way it seemed like a good neutral point.

Reply:
Personally, I think the use of macros for repeated code is fine, since Rust macros don't really introduce new potential bugs the way that, say, the C preprocessor does. I'm not _convinced_ they're strictly necessary, since the code in question is repeated twice, but I do think it's good to factor it out, so that any future changes can be made in one place rather than two.

The main thing I'd change is that I don't _really_ understand why it's necessary to have two separate macros --- it looks like the output of `tap_service!` is only ever passed directly into `tap_task!`. It seems like we could combine them into one macro?

Reply:
They were separated because of type inference errors that occur due to the difference in how we create `srv` between the two call sites--the `map_err` and `and_then` in the accept branch specifically.

Reply:
I'm more concerned about readability than introducing bugs

Reply:
As functions they are no more readable than macros unfortunately. The signatures--similar to `serve_tap`--are longer than the actual bodies. I can remove them to help with readability.

Reply:
Just so I understand... the reason why this can't be a function is because the return type is some horrible insane type?  And if we just shove that in a let binding, type inference will just silently know what type it is.  But if we make this into a function then suddenly we have to write down the name of that type and our eyeballs will melt out of our faces?

And you can't have a function with an inferred return type?

Rust is weird, y'all.

Reply:
@kleimkuhler I took a crack at replacing both macros with a single function. Here's the diff: https://github.com/linkerd/linkerd2-proxy/compare/kleimkuhler/tap-identity-chk...eliza/kleimkuhler/tap-identity-chk?expand=1#diff-477a4e74a1956a896487a859ef4fc481

If you're interested, I can open a PR against this branch to merge those changes.

Reply:
@hawkw This is awesome thanks! It is merged now